File: leetcode_python/42.py
from typing import List
class Solution:
    # 错误的解法
    # def trap(self, height: List[int]) -> int:
    #     l  = len(height)
    #     if l == 0:
    #         return 0
    #     left , right = 0, 0
    #     while  left<l and height[left]==0 :
    #         left += 1
    #     if left==l:
    #         return 0
    #     right = left+1
    #     ans = 0
    #     while right<l and left<l:
    #         while  right<l and height[right]<height[left] :
    #             right += 1
    #         if right==l:
    #             left +=1
    #             right = left+1
    #         else:
    #             ans += min(height[left],height[right])*(right-left-1)-sum(height[left+1:right])
    #             left = right
    #             right = left+1
    #     return ans
    # 暴力方法 超出时间限制 
    def trap_violent_solution(self, height: List[int]) -> int:
        l = len(height)
        if l <= 0:
            return 0
        ans = 0
        for i in range(1,l-1):
            max_left , max_right = 0 ,0 
            # 向左边探索
            for a in range(i,-1,-1):
                max_left = max(max_left,height[a])
            # 向右边探索
            for b in range(i,l):
                max_right = max(max_right,height[b])
            ans += min(max_left,max_right)-height[i]
        return ans
    # 使用动态规划
    def trap(self, height: List[int]) -> int:
        pass


s = Solution()
print(s.trap([4,2,0,3,2,5]))

# 42. 接雨水
# https://leetcode-cn.com/problems/trapping-rain-water/
from typing import List
from collections import defaultdict
from collections import deque
from math import inf
import bisect
import heapq
'''
@File : 42.py
@Time : 2022/01/25 17:58:03
@Author : YuMin Zhang
@State : Half-Depedent
@Thinking :
@Tag : Hard
'''
class Solution:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0
        l = len(height)
        if l <= 2:
            return 0
        left_high = [0] * l 
        left_high[0] = height[0]
        tmp_left = height[0]

        for i in range(1, len(height)):
            left_high[i] = tmp_left
            tmp_left = max(tmp_left, height[i])
        
        right_high = [0] * l
        tmp_right = height[-1]
        right_high[-1] = height[-1]
        for i in range(len(height) - 2, -1, -1):
            right_high[i] = tmp_right
            tmp_right = max(tmp_right, height[i])
        ans = 0
        for i in range(1, l):
            min_height = min(left_high[i], right_high[i])
            if min_height > height[i]:
                ans += min_height - height[i]
        return ans

            

      

File: leetcode_c_plus_plus/42.cpp
// 42. 接雨水
// https://leetcode-cn.com/problems/trapping-rain-water/
#include<map>
#include<set>
#include<vector>
#include<cmath>
#include<numeric>
#include<unordered_map>
using namespace std;
#define INT_MAX 2147483647
#define INT_MIN (-INT_MAX - 1)
/**
@File : 42.cpp
@Time : 2022/04/07 20:32:39
@Author : YuMin Zhang
@State : Indepeedent | Half-Depedent | Depedent 
@Thinking :
@Tag : Easy | Medium | Hard
**/
class Solution {
public:
    int trap(vector<int>& height) {
        int ans = 0;
        int l = height.size();
        int left = 0;
        int right = 0;
        int h = 1;
        int max_val = 0;
        for(auto x : height) {
            max_val = max(max_val, x);
        }
        while(h <= max_val) {
            for(int i = 0; i < l; i++) {
                if(height[i] >= h) {
                    left = i;
                    break;
                }
            }
            right = left + 1;
            while(right < l) {
                if(height[right] >= h) {
                    ans += (right - left - 1);
                    left = right;
                }
                right++;
            }
            h++;
        }
        return ans;
    }
};

class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        if (n == 0) {
            return 0;
        }
        vector<int> leftMax(n);
        leftMax[0] = height[0];
        for (int i = 1; i < n; ++i) {
            leftMax[i] = max(leftMax[i - 1], height[i]);
        }

        vector<int> rightMax(n);
        rightMax[n - 1] = height[n - 1];
        for (int i = n - 2; i >= 0; --i) {
            rightMax[i] = max(rightMax[i + 1], height[i]);
        }

        int ans = 0;
        for (int i = 0; i < n; ++i) {
            ans += min(leftMax[i], rightMax[i]) - height[i];
        }
        return ans;
    }
};